src/v/datalake/event_multiplexer.h

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+#pragma once
+
+#include "container/fragmented_vector.h"
+#include "datalake/schemaless_translator.h"
+#include "json/writer.h"
+#include "model/record.h"
+
+// TODO Only needed for method definitions, not class. Move methods out of line
+// if possible
+#include "storage/parser_utils.h"
+
+#include <seastar/core/future.hh>
+
+#include <optional>
+
+namespace datalake {
+/*
+Consumes logs and sends events to the appropriate translator
+based on the schema ID. This is meant to be called with a
+read_committed_reader created from a kafka::partition_proxy.
+
+This is a skeleton class. Currently it only uses the trivial
+schemaless_translator.
+
+pseudocode:
+for each event {
+    t = get_translator(event.schema_id);
+    d = t.translate(event);
+    w = get_data_writer(event.schema_id);
+    w.add_event(d);
+}
+
+The WriterFactory is a class with
+- instance method `create_writer`
+- nested class `data_writer_result`
+- nested class `data_writer`
+  - instance method `add_data_struct` to consume a parsed event
+  - instance method `finish` to clean up and return a `data_writer_result`
```

Comment:
nit: it would be great to encode these as a concept. Or use a base class and heap allocate? Is this only configurable for testing?